### ISSUE TITLE
fifo_queue consumers can stuck forever.

### ISSUE THREAD

**Issue by absorbb** [enriquebris/goconcurrentqueue#30]
Hi!
It looks like fifo_queue is not really safe to use and can lock the app forever.

Lets say we have length=0 
and consumer calls for  DequeueOrWaitForNextElementContext
and reaches line 117

Meanwhile producer tries to Enqueue a new object:
It takes `listener` in line 46.
but can't send object yet in  line 49 because no one listen yet 
so it adds to the slice and forgets about `listener`

Consumer start to consume `waitChan` in 126 but that will never happen .
i guess the whole idea of the `for` loop is to resolve that. But there is no number of iterations and timeout values that can guarantee that it will be able to Dequeue anything in 129. Other consumers can always get there faster.
So loop ends with no results and we end up trying to consume `waitChan` forever in 137

It is hard to catch but very painful.
Number of consumers > 1 required to catch that.

**Comment by enriquebris** [comment]
Hi @absorbb !

Please confirm you are using the [latest version: v0.6.1](https://github.com/enriquebris/goconcurrentqueue/releases/tag/v0.6.1). I just created this tag today because *go get* wasn't pulling the latest code.

Could you share any code snippet experiencing this issue? It would help, for sure.

I'm going to try to recreate the exact scenario you explained above. I'll let you know as soon as I get something.

**Comment by absorbb** [comment]
Yeah. I've used the latest code. We have actually copied .go file into our project.

Snippet:
```go
package main

import (
	"fmt"
	"github.com/enriquebris/goconcurrentqueue"
	"math/rand"
	"time"
)

func main() {
	rand.Seed(time.Now().Unix())
	fifo := goconcurrentqueue.NewFIFO()
	for i := 0; i < 11; i++ {
		producerNumber := i
		go func() {
			for {
				fifo.Enqueue(fmt.Sprintf("(producer: %d)", producerNumber))
                                //some unpredictable pauses to simulate real life events flow
				toggle := rand.Intn(2)
				if toggle == 0 {
					time.Sleep(time.Millisecond * time.Duration(rand.Intn(2000)))
				}
			}
		}()
	}
	for i := 0; i < 10; i++ {
		consumerNumber := i
		go func() {
			for {
				msg, err := fifo.DequeueOrWaitForNextElement()
				if err != nil {
					fmt.Println(err)
				} else {
					fmt.Printf("Consumer %d: message: %s\n", consumerNumber, msg)
				}
			}
		}()
	}
	for {
		time.Sleep(time.Second)
	}
}
```
Run it and eventually all consumers will stuck. For me first consumer get stuck quite quickly.
And took about 20 minutes for all other to get there on Apple M1 processor. So the last one get stuck too. that means that my statement that "Number of consumers > 1 required to catch that." wasn't correct.

I believe that Fixed FIFO is affected too.

**Comment by enriquebris** [comment]
Ok, got the idea, thanks for the code snippet!

I'm working in order to fix this particular issue. Good catch!

**Comment by enriquebris** [comment]
Hi @absorbb,

The issue you reported is currently fixed in [v0.6.2](https://github.com/enriquebris/goconcurrentqueue/releases/tag/v0.6.2):

```bash
go get -u github.com/enriquebris/goconcurrentqueue@v0.6.2
```

Please confirm this new version fits your project.

**Comment by absorbb** [comment]
This version produces start spamming
`empty queue and can't wait for next element because there are too many DequeueOrWaitForNextElement() waiting
`
error after few minutes of running my code snippet.

**Comment by absorbb** [comment]
@enriquebris problem is that snippet starts only 10 consumers. And on my run i used even less - 2 consumers.

**Comment by enriquebris** [comment]
ok, looking into it

**Comment by enriquebris** [comment]
Hey @absorbb,

The original issue was solved by [v0.6.2](https://github.com/enriquebris/goconcurrentqueue/releases/tag/v0.6.2), but it also added a scenario in which some useless callback channels were enqueued by FIFO.DequeueOrWaitForNextElementContext after many executions. This second scenario is currently solved by [v0.6.3](https://github.com/enriquebris/goconcurrentqueue/releases/tag/v0.6.3):

```bash
go get github.com/enriquebris/goconcurrentqueue@v0.6.3
```

Please confirm this new version fits your project.

**Comment by enriquebris** [comment]
Closing this issue as it was solved by [v0.6.3](https://github.com/enriquebris/goconcurrentqueue/releases/tag/v0.6.3).
